Dev contest to rapidly switch between contest status

rissois opened the pull request:
Requires database update:
`ALTER TABLE entries ADD background_color enum('#000000', '#FFFFFF', '#4b91e3');`
`ALTER TABLE votes ADD is_desktop boolean;`
`ALTER TABLE votes ADD is_keyed boolean;`

heshammourad commented:
It's probably too late for this pull request, but there's too much going on in a single pull request. There are at least three separate efforts: DevBar, background color, and what appear to be general refactoring for data and voting. It's a bit difficult to track what each change is related to and context switch between the concerns. If possible, could you break this up into separate PRs or at least keep that in mind for future ones?

heshammourad commented on react-ui/src/utils/useSwrContest.jsx:
```diff
@@ -0,0 +1,28 @@
+  const contestId = overrideId || useContestId();
+  const apiPath = `/contests/${contestId}`;
+
+  const { data = {}, error: e, ...rest } = useSwrAuth(apiPath, { ...revalidateOnMount });
```
What does it mean to use the spread operator on a `bool`?

rissois replied:
It doesn't mean anything. I think I had converted to `options` at once point, then sloppily reverted back. Fixed now.

heshammourad commented on react-ui/src/utils/useSwrAuth.jsx:
Maybe all of the `useSwrXyz` hooks can go in a `data` directory? We could rename `api` to `data` and put all of these in there.

rissois replied:
Ya, that sounds really clean actually. Will do that once all else is cleared, let's leave this convo as unresolved in the meantime.

heshammourad replied:
Resolved in #294.

heshammourad commented on server/api/dev.js:
```diff
@@ -0,0 +1,214 @@
+/*
+  Easier to refresh entire dev entry instead of upsert/update the dates
+  ??? What is the removed column for in entries?
```
No one knows 😅 

https://github.com/heshammourad/vexillology-contests/wiki/Database#entries

rissois replied:
lol, ok

rissois commented:
My bad, I didn't realize PRs couldn't be worked through commit-by-commit, let me know the best way to do this next time. What always happens is I start on feature X, but then I realize it would help to do Y and Z.

heshammourad commented on server/api/dev.js:
```diff
@@ -0,0 +1,214 @@
+  const pastest = new Date(NOW.getTime() - fiveYears - 1).toISOString();
+  const past = new Date(NOW.getTime() - fiveYears).toISOString();
+  const future = new Date(NOW.getTime() + fiveYears).toISOString();
+  const futurest = new Date(NOW.getTime() + fiveYears + 1).toISOString();
```
[date-fns](https://date-fns.org/docs/Getting-Started) is what I've been using as my date time library. It's a little more readable than this and we don't have to do calculations like `1000 * 60 * 60 * 24 * 365 * 5` (which could be a class CONST instead of defined in the function).

```javascript
const yearEnd = format(NOW, 'MM') === 12; // or format(NOW, 'MMM') === 'Dec'
const date = formatISO(NOW, { representation: 'date' });
const pastest = sub(NOW, { years: 5, months: 1 });
const past = sub(NOW, { years: 5 });
const future = add(NOW, { years: 5 });
const futurest = add(NOW, { years: 5, months: 1 });
```

rissois replied:
Man, the Postgres timestamp was killing me. Used what you said above, but I actually needed to add `.toISOString()` to get this to work. You're right though, looks way better.

heshammourad commented on server/api/dev.js:
```diff
@@ -0,0 +1,214 @@
+const db = require('../db');
+const { createLogger } = require('../logger');
+
+const NOW = new Date();
```
Do you know if this stays constant for the entire run of the server since it's defined outside of the function? Or is instantiated again every time a function in the file is called? This sounds like one of those interview questions, and I'm not sure what the answer is.

rissois replied:
Good catch! It does stay the same. I fixed this.

heshammourad commented on server/api/dev.js:
```diff
@@ -0,0 +1,214 @@
+  const contestIndex = contestStatuses.indexOf(status);
+  const isAfter = (lastStatus) => contestIndex > contestStatuses.indexOf(lastStatus);
+
+  const yearEnd = NOW.getMonth() === 11; // getMonth is zero-based
```
FYI, there are actually two contests in December, a regular one and a year-one end. Not sure if this affects anything here, though.

rissois replied:
This should not impact the `dev` contest

heshammourad commented on server/api/dev.js:
```diff
@@ -0,0 +1,214 @@
+      */
+      await db.del('votes', { contest_id: 'dev' });
+      await db.del('contest_entries', { contest_id: 'dev' });
+      await db.any("DELETE FROM entries WHERE id LIKE 'dev%';");
```
Couldn't there possibly be real IDs that start with `'dev'`? There's no special characters in the UUID alphabet we're using, i.e. `[A-Za-z0-9]`, so maybe we should add a special character in the id like `dev-`.

rissois replied:
Good catch, used the hyphen as you suggested

heshammourad commented on server/api/dev.js:
```diff
@@ -0,0 +1,214 @@
+        ------- CONTEST_CATEGORIES -------
+        Create categories for dev contest
+      */
+      const categories = [{ contest_id: 'dev', category: 'cat1' }, { contest_id: 'dev', category: 'cat2' }];
```
Maybe something for the future, but it is sometimes helpful to test to see what submission and contests look like without categories.

rissois replied:
Ya, the exact content of `dev.js` `entryVersions` will need some tweaking with time

heshammourad commented on server/api/dev.js:
```diff
@@ -0,0 +1,214 @@
+        users.push({ username });
+      }
+      if (!users.some((user) => user.username === 'dev')) {
+        await db.insert('users', [{ username: 'dev', contest_reminders: true, moderator: false }]);
```
FYI, only moderators are required to be in the database since the modified_by (for submission status) is the only field that refers to the users table. The regular username column is just a string column with no dependencies.

rissois replied:
Ah ok. Do you want me to remove it, or just keep it in in case the keys change in the future?

heshammourad replied:
I think it's fine to keep it, just something that I figured wasn't entirely intuitive about the database and wanted to point out.

heshammourad commented on server/api/dev.js:
```diff
@@ -0,0 +1,214 @@
+      /*
+        ------- ENTRIES -------
+        Create some flags for yourself and dev user
+        Adust as desired
```
Nit: adjust

heshammourad commented on server/api/dev.js:
```diff
@@ -0,0 +1,214 @@
+      // add sequential dev# for id (NOTE: zero-indexed, can do i+1 if desired)
+      entryVersions = entryVersions.map((v, i) => ({ ...v, id: `dev${i}`, background_color: BACKGROUND_COLORS[i % 3] }));
+
+      const existingEntries = await db.select(`SELECT width, height, url FROM entries WHERE height > 600 AND POSITION('https://firebasestorage.googleapis.com' IN url) > 0 LIMIT ${entryVersions.length}`);
```
Just curious, but why not `AND url LIKE '%firebasestorage%'`?

rissois replied:
Originally had that, sort of forget why I thought using the entire url felt safer but switched to this

heshammourad commented on server/api/dev.js:
```diff
@@ -0,0 +1,214 @@
+      const contest = getDevContest(status);
+      // Switch contest status
+      if (isExisting) {
+        // update rejects date for some reason
```
I think you need to cast the date in to a string for it to work.

rissois replied:
Idk, `toISOString()` still doesn't pass here.

heshammourad replied:
From what I can tell ([StackOverflow](https://stackoverflow.com/questions/75485625/is-it-possible-to-enum-and-timestamp-using-pgp-helpers-update)), you need to define the cast in your column set. So it could be something like:
```javascript
await db.update('contests', [contest], Object.entries(contest).map(([key, value]) => {
  if (key === 'id') {
    return '?id';
  }
  if (value instanceof Date) {
    return { name: key, cast: 'timestamp' };
  }
  return key;
});
```

heshammourad commented on server/api/dev.js:
```diff
@@ -0,0 +1,214 @@
+        // update rejects date for some reason
+        // await db.update('contests',
+        // [contest], Object.keys(contest).map((key) => (key === 'id' ? '?id' : key)));
+        await db.any(`UPDATE contests
+          SET
+            name = '${contest.name}',
+            date = '${contest.date}',
+            year_end = ${contest.year_end},
+            submission_start = '${contest.submission_start}',
+            submission_end = '${contest.submission_end}',
+            vote_start = '${contest.vote_start}',
+            vote_end = '${contest.vote_end}'
+          WHERE id = 'dev';`);
```
Looks like this caused a code scanning alert: `Database query built from user-controlled sources`. Using `db.update` should avoid this.

rissois replied:
Can you play around with update on your end? I just couldn't get past the update issue.

heshammourad replied:
Sure, I realized there isn't an update that currently puts a date, so I was just speculating.

heshammourad commented on server/api/submission.js:
```diff
@@ -110,6 +114,7 @@ exports.post = async (
       width,
       url,
       user: username,
+      background_color: backgroundColor,
```
Nit, would it be possible to keep object keys in alphabetical order when you are adding them? It would be great to have a linter do this, but in the absence of that availability, it improves readability and being able to find/search for specific keys later when there is a semantic order.

rissois replied:
Ya, I try to catch these because I never got the linter to behave.

heshammourad replied:
Yeah, it's a bit inconsistent in how they work. Speaking of which, eslint is deprecating formatting: https://eslint.org/blog/2023/10/deprecating-formatting-rules/, so we might have to tweak our prettier and eslint configs to account for that some time in the future.

heshammourad commented on server/db/index.js:
```diff
@@ -97,9 +97,23 @@ const update = async (table, data, columns, returning = []) => {
+//     DO UPDATE SET ${set};
+//   `);
+// };
+
```
Remove commented out code. (Or uncomment and keep because this looks like a helpful function to have)

rissois replied:
Created issue  #305

heshammourad commented on react-ui/src/utils/useStorage.jsx:
Do we need the empty file?

rissois replied:
🤦

rissois commented on react-ui/src/components/DevBar.jsx:
```diff
@@ -0,0 +1,98 @@
+/**
+ * Banner to notify developer of overrides in env.js
+ * ??? ideally this would be top of document, but fighting with AppBar/Toolbar
```
Ya, that was my original thought, but was fighting with it hiding the ContestSponsor. Figured out a solution finally.

rissois commented on react-ui/src/components/DevBar.jsx:
```diff
@@ -0,0 +1,98 @@
+  const name = error?.data?.name || data?.name;
+  const { data: { moderator: isModerator }, mutate: mutateMod } = useSwrInit();
+  const [{ isLoggedIn, accessToken, refreshToken }] = useAuthState();
+  const authTokens = { accessToken, refreshToken };
```
_IMPORTANT_

Done, but make sure you log out and log back on for AuthorizeCallback to fire and create the authTokens variable

heshammourad replied:
Ah, good catch. Will that affect everyone currently logged into the site?

rissois replied:
Shoot, I forgot about everyone else lol. Force log out is easiest, but we don't want to do that too often and I'd rather save it for a bigger reason.

Instead, I rewrote `useSwrAuth` as:
```
  const [{
    refreshTokens, // Delete in future force quit
    accessTokens, // Delete in future force quit
    authTokens = { refreshTokens, accessTokens },
  }] = useAuthState();
```

rissois commented on react-ui/src/pages/entry/EntryModal.jsx:
```diff
@@ -14,18 +14,20 @@ import {
   PageWithDrawer,
   RedditLogInDialog,
 } from '../../components';
+// eslint-disable-next-line import/no-unresolved
```
I don't know actually, when I deleted the comment the problem no longer existed.

rissois commented on react-ui/src/pages/entry/EntryModal.jsx:
```diff
@@ -160,6 +162,10 @@ function EntryModal() {
     let indexChange = 0;
     if (key >= '0' && key <= '5') {
+      // ??? I would love it if we could track whether votes mobile or desktop
+      // ??? and if desktop, how many were keyed
```
One option we could do for now is just create two columns on votes: is_desktop and is_keyed. Then we can create a variable in env.js for IS_DESKTOP where width > 800 (or something).

It's not full analytics, but I'd love to see 1. number of mobile voters and 2. adoption of the keyed vote so I can decide if it is worth designing a quick-vote system on mobile.

Lmk what you think.

heshammourad replied:
That's not a bad idea, gets an idea of how useful it is until we can implement proper analytics.

rissois replied:
Ok, I added this implementation. I followed MDN recommendations and code to check isTouchScreen instead of width. Works well, but can't test on mobile.

heshammourad replied:
Neat, that'll probably pick some tablet-like devices like Chromebooks that also have touchscreens, but I don't think we get a lot of usage on those devices. I think this approach will be a reasonable estimate.

We can test once we merge this PR and it's available in staging.

rissois commented on react-ui/src/pages/hallOfFame/HallOfFameCard.jsx:
```diff
@@ -101,12 +102,14 @@ function HallOfFameCard({
-        image={imagePath}
-        width={width}
-      />
+      <CardActionArea>
```
I believe `CardActionArea` predates me

heshammourad replied:
Looks like `CardActionArea` isn't present in the main branch, so I don't think it needs to be added here.

rissois replied:
Ah, now I remember why this was hear. For some reason, the `Skeleton` will all bunch up in the top-left corner of the page without `CardActionArea`. Copying from the Contest was the simplest way I could find to make `Skeleton` display in-place.

I added `disabled` so the flag is not clickable instead.

heshammourad replied:
Could this just be because `CardActionArea` adds a `div` around its content? Would adding a `div` or `Box` work here instead?

rissois replied:
Strangely no, only `CardActionArea` works.

heshammourad replied:
🤷 Oh well, could you please add a `// TODO` to look at it "later"?

rissois replied:
Good catch, should be good now

rissois commented on react-ui/src/pages/submission/Submission.jsx:
```diff
@@ -135,277 +38,48 @@ const useStyles = makeStyles((theme) => ({
-      categories,
-      firebaseToken,
+      // categories,
+      // firebaseToken,
```
Deleted, not sure why they were left in

rissois commented on react-ui/src/pages/submission/SubmissionPrompt.jsx:
```diff
@@ -0,0 +1,32 @@
+  RedditMarkdown,
+} from '../../components';
+
+import May23 from './content/May23';
```
I think we will want an `Accordion` in the future that can show visitors what the old prompts were, and not just the competitions. I created an issue #280 for this.

rissois commented on react-ui/src/utils/useContestId.jsx:
I sort of just needed a separate folder to work in, it was just easier to stay organized as I migrated a few concepts over. Happy to re-add them back to `/common/` once we are confident that `utils` are working correctly

heshammourad replied:
Sounds good, I think one `common` or `utils` folder with everything in it probably makes more sense, unless there's a better breakdown you can think of. Maybe hooks be in a separate folder? Not sure.

heshammourad commented:
> My bad, I didn't realize PRs couldn't be worked through commit-by-commit, let me know the best way to do this next time. What always happens is I start on feature X, but then I realize it would help to do Y and Z.

Yeah, it's a little difficult when some of the functionality is shared. Ideally, you could create separate branches for each distinct effort and then create a PR for each branch. It's not the easiest to do when you have functionality that depends on functionality you did in another branch.

You could also create branches off of branches instead of off of main and then create PRs to merge the feature branches into each other. Would probably be more merge conflicts to deal with.

heshammourad commented on server/api/submission.js:
```diff
@@ -71,7 +75,7 @@ const isWithinSubmissionWindow = async () => {
   {
     body: {
-      category, contestId, description, height, name, url, width,
+      category, contestId, description, height, name, url, width, backgroundColor,
```
One thing I thought of, should we add a validation for background color that it's one of the 3 values we have? Don't want someone submitting (e.g using postman) a value that is either a weird color or worse isn't a color.

rissois replied:
Are you thinking:
`ALTER TABLE entries ADD background_color enum('#000', '#FFF', '#4b91e3');`

We can also do full hex to be consistent

heshammourad replied:
Yeah, if it makes it easier to do a regex for it.

Another (probably a bit overengineered) is to have a `background_colors` table, so the `entries` table only accepts colors we have in there. Those colors could then be passed down on the `/api/submissions` respone so we wouldn't have to update the front-end and back-end with new values.

rissois replied:
Ok, switched to full hex just for consistency.

rissois commented on react-ui/src/pages/contest/Contest.jsx:
```diff
@@ -54,24 +51,22 @@ const useStyles = makeStyles((theme) => ({
+  if (contest?.submissionWindowOpen) {
+    navigate('/submission', { replace: true });
+  }
+  // }, [contest.submissionWindowOpen]);
```
This was the offending code:
```
useEffect(() => {
    forceCheck();
    window.history.pushState(
      { usr: { ...window.history.state?.usr, selectedCategories } },
      document.title,
    );
  }, [selectedCategories]);
  ```

I changed to `replaceState`, which fixes the issue. What is `usr` for anyways?

heshammourad replied:
I can't remember, I think I needed to transfer `selectedCategories` state between Contest and Entry. It shouldn't be necessary any more now that Entry is a modal.

rissois replied:
You're right, just deleted entire block instead.

rissois commented on react-ui/src/utils/LocalStorageProvider.jsx:
```diff
@@ -0,0 +1,26 @@
+import { START_WITHOUT_CACHE } from "../env";
+
+
+export default function localStorageProvider() {
```
I _believe_ this only caches swr data atm. `useSwrData` was the only place where `useCache` was used. I did not mess around with `use-persisted-state` yet, so `useComponentsState` should be unchanged.

rissois commented on react-ui/src/pages/submission/SubmissionForm.jsx:
```diff
@@ -0,0 +1,619 @@
+ * @param {func} setPreviewDescription - Function to change previewDescription
+ * @param {func} setSelectedTab - Go to different submission tab
+ * @param {bool} submissionExpired - is submission window open
+ */
```
Done

rissois commented:
I believe I covered everything in the latest commit except:
- Cannot test `is_keyed` and `is_desktop` voting on mobile
- `CardAreaAction` appears to be only solution to `HallOfFame` at the moment
- Folders `api` and `utils` will be merged into one folder `data`, though some will be moved into `common`
- Postgres date / timestamp issue and update (if this is fixed, the commented out upsert code might actually work) 

Let me know if we need a separate issue for `use-persisted-state`.

I added the database alter commands at the top of the PR, but will copy them here:
```
ALTER TABLE entries ADD background_color enum('#000000', '#FFFFFF', '#4b91e3');
ALTER TABLE votes ADD is_desktop boolean;
ALTER TABLE votes ADD is_keyed boolean;
```

heshammourad commented:
Awesome, I'll check it out and try running the code. I'll also play around with the `update` query and see what's needed there.

If there isn't an issue for replacing `use-persisted-state`, please add one. It's definitely needed because the package isn't being maintained, and it's also incompatible with react-18 which I eventually want to migrate to.

heshammourad commented:
I'm starting to look into this today, but one thing I was thinking. What about adding an analytics table instead of putting those columns in votes? We could just add any events we can think of to the table:
```
event ENUM(VOTE_DESKTOP, VOTE_MOBILE, VOTE_KEY)
time TIMESTAMP_TZ
```

rissois commented on react-ui/src/components/PageWithDrawer.jsx:
```diff
@@ -81,6 +82,7 @@ function PageWithDrawer({
   appBar, children, className, drawer, handleClose, isOpen,
 }) {
   const classes = useStyles();
+  const isModal = appBar.accountMenuColor === 'inherit';
```
Ya, was not a very clean solution. I just added an isModal prop instead.

rissois commented on react-ui/src/components/DevBar.jsx:
Fixed

rissois commented on server/api/contest.js:
```diff
@@ -181,7 +180,7 @@ exports.get = async ({ params: { id }, username }, res) => {
               'rank',
             ]);
-            await db.update('entries', entriesData, ['?id', 'description', 'name', 'user']);
+            await db.update('entries', [entriesData], ['?id', 'description', 'name', 'user']);
```
Don't know why that got changed, reversed

rissois commented on react-ui/src/components/AccountMenu.jsx:
```diff
@@ -50,12 +51,12 @@ function AccountMenu({ color }) {
   const classes = useStyles();
 
-  const [{ isLoggedIn, refreshToken, username }, setAuthState] = useAuthState();
+  const [{ isLoggedIn, authTokens: { refreshToken }, username }, setAuthState] = useAuthState();
```
I think option 3 is most appropriate, especially in case we implement useAuthState elsewhere. Made the edits.

rissois commented:
> I'm starting to look into this today, but one thing I was thinking. What about adding an analytics table instead of putting those columns in votes? We could just add any events we can think of to the table:
> 
> ```
> event ENUM(VOTE_DESKTOP, VOTE_MOBILE, VOTE_KEY)
> time TIMESTAMP_TZ
> ```

I think this is something that should be a separate PR once we can discuss the events and fields required for an analytics table.

rissois commented:
For some reason I can't reply directly to your suggestion for casting the timestamps, but the code didn't work. I think the update code recasts it as a string maybe?

heshammourad commented:
> For some reason I can't reply directly to your suggestion for casting the timestamps, but the code didn't work. I think the update code recasts it as a string maybe?

Could you send the code snippet for this? Also, if you set `LOG=debug` in your .env file it should print the generated query to the console. Can you send what it's outputting for this?

heshammourad commented:
There's several eslint warnings in the console:
```
src/common/index.js
  Line 10:1:  Dependency cycle via ../utils/useSwrInit:8=>./useSwrAuth:10                           import/no-cycle
  Line 16:1:  Dependency cycle via ./useVotingStatus:16=>../utils/useSwrContest:11=>./useSwrAuth:8  import/no-cycle
  Line 17:1:  Dependency cycle via ../utils/useSwrContest:11=>./useSwrAuth:8                        import/no-cycle

src/common/useAuthState.jsx
  Line 8:1:   Expected indentation of 2 spaces but found 4  indent
  Line 10:1:  Expected indentation of 2 spaces but found 4  indent
  Line 11:1:  Expected indentation of 2 spaces but found 4  indent
  Line 12:1:  Expected indentation of 2 spaces but found 4  indent
  Line 13:1:  Expected indentation of 4 spaces but found 8  indent
  Line 14:1:  Expected indentation of 2 spaces but found 4  indent
  Line 16:1:  Expected indentation of 2 spaces but found 4  indent

src/common/useVotingStatus.jsx
  Line 32:95:  Missing semicolon  semi

src/components/EntryDescriptionDrawer.jsx
  Line 37:3:  Class `banner` is unused  mui-unused-classes/unused-classes

src/env.js
  Line 23:1:  Expected indentation of 2 spaces but found 4  indent
  Line 24:1:  Expected indentation of 2 spaces but found 4  indent
  Line 25:1:  Expected indentation of 2 spaces but found 4  indent

src/utils/LocalStorageProvider.jsx
  Line 7:37:  Strings must use singlequote                  quotes
  Line 9:1:   More than 1 blank line not allowed            no-multiple-empty-lines
  Line 11:1:  Expected indentation of 2 spaces but found 4  indent
  Line 12:1:  Expected indentation of 4 spaces but found 8  indent
  Line 13:1:  Expected indentation of 2 spaces but found 4  indent
  Line 15:1:  Expected indentation of 2 spaces but found 4  indent
  Line 16:1:  Expected indentation of 2 spaces but found 4  indent
  Line 18:1:  Expected indentation of 2 spaces but found 4  indent
  Line 19:1:  Expected indentation of 2 spaces but found 4  indent
  Line 20:1:  Expected indentation of 4 spaces but found 8  indent
  Line 21:1:  Expected indentation of 4 spaces but found 8  indent
  Line 22:1:  Expected indentation of 2 spaces but found 4  indent
  Line 24:1:  Expected indentation of 2 spaces but found 4  indent
  Line 25:1:  Expected indentation of 2 spaces but found 4  indent
```

heshammourad commented on react-ui/src/pages/entry/EntryModal.jsx:
```diff
@@ -160,6 +161,11 @@ function EntryModal() {
+      // https://stackoverflow.com/questions/39435395/reactjs-how-to-determine-if-the-application-is-being-viewed-on-mobile-or-deskto
+      changeRating(parseInt(key, 10), entry?.rating, true);
+      indexChange = 1;
       return;
```
By returning here, the `indexChange` never gets used. Perhaps wrap the `switch` in an `else` clause and delete the `return`.

Also, is there a scenario where someone is working backwards for some reason and ends up having to click back twice? Maybe we should add an issue to change this setting, i.e. have auto-next be the default but at least allow the user to toggle it off?

rissois replied:
I moved the if statement into the switch default because it seemed a bit cleaner than the else. Another option is to just add `handleNavigate` into the `if` statement before the `return`.

I created an issue, but I'd imaging >99% of users will follow forward progression, and if you are going backwards than double-clicking back is manageable.

heshammourad replied:
Yeah, putting it in the default case makes sense.

And agreed, it's a low priority issue, but maybe we add it sometime down the line.

heshammourad commented on react-ui/src/pages/entry/EntryModal.jsx:
```diff
@@ -122,8 +116,15 @@ function EntryModal() {
+  /**
+   * Navigation and voting
    */
+  const { changeRating, clearRating } = useVoting(entryId);
```
‼️ MAJOR BUG ‼️ 

`useVoting` is only instantiated once when the component is rendered, and these functions always apply to the `entryId` of the first entry loaded.

rissois replied:
Good catch, I hadn't caught that `handleKeyUp` was in a useEffect. I added it to the dependencies array.

rissois commented on server/env.js:
```diff
@@ -1,25 +1,35 @@
+ */
+
+const ALLOW_UNAUTHENTICATED = false;
+const CURRENT_CONTEST_DEV = true;
```
Changed to `&& IS_DEV` for all env.js files. There are actually three options here:
1. `const ALLOW_UNAUTHENTICATED = false;` and `const IS_UNAUTHENTICATED_VIEW = IS_DEV && ALLOW_UNAUTHENTICATED`
2. `const ALLOW_UNAUTHENTICATED = false && IS_DEV;`
3. `const ALLOW_UNAUTHENTICATED = false;` and `module.exports = {ALLOW_UNAUTHENTICATED: IS_DEV && ALLOW_UNAUTHENTICATED}`

I initially thought 1 was cleanest, but realized all the extra variables were going to pile up. 3 is another option if we want to initialization to look cleaner, while 2 lets you know this is dev only.

heshammourad replied:
Wouldn't 2 always evaluate to `false`?

Here's a (probably convoluted) way to ensure that all env variables only intended for use in a DEV environment always take `IS_DEV` into account.

```javascript
const DEV_ENV = {
  ALLOW_UNAUTHENTICATED: false,
  CURRENT_CONTEST_DEV: false,
};

const DEV_ENV_EXPORT = Object.entries(DEV_ENV).reduce((acc, [key, value]) => {
  acc[key] = value && IS_DEV;
}, {});

module.exports = {
  ...DEV_ENV_EXPORT,
};
```

rissois replied:
Sorry, each `false` in my previous comment was meant to be a user-controlled variable. But I'll implement the code above instead, looks cleaner.

rissois replied:
Is there a way to have a default value whenever I push to Github? I don't want to accidentally turn on `START_WITHOUT_CACHE`, for instance.

heshammourad replied:
Only the .env file has server defaults. I am not aware of a way to put a check on JS file contents, beyond maybe manually adding some comment like `DO_NOT_SUBMIT` so it's clear.

heshammourad replied:
Anything that changes prod behavior should be defaulted to `false` in the committed version and changed only in the local version of each developer. I was running into it trying to pull the non-existent 'dev' contest.

rissois commented on react-ui/src/pages/submission/Submission.jsx:
```diff
@@ -135,277 +38,46 @@ const useStyles = makeStyles((theme) => ({
+
+  const noData = (isLoading
+    ? <CircularProgress />
+    : <div>The submission portal is not yet open. Please check again later.</div>);
```
Sorry, noData was meant to take the place of 404. I raised all the not open and closed divs.

rissois commented on react-ui/src/pages/submission/Submission.jsx:
```diff
@@ -135,277 +38,46 @@ const useStyles = makeStyles((theme) => ({
-  const updateSnackbarState = useSnackbarState();
   const [selectedTab, setSelectedTab] = useState(state?.defaultTab ?? 0);
-  const [fileDimensions, setFileDimensions] = useState(null);
   const [previewDescription, setPreviewDescription] = useState(false);
```
I left it as required since `previewDescription` comes from `useState`, so if that's missing we've got a deeper issue. I did fix TextField, JSDoc, and propType though.

Next project will definitely be Typescript

heshammourad replied:
`isRequired` means the value provided is non-null and not undefined. If `useState`'s default is `null` this will fail the prop check in SubmissionForm.

Agreed that Typescript should make things easier. It should theoretically be possible to make new files in TSX while still having JSX, seems like it's just adding a `tsconfig.json`: https://www.totaltypescript.com/add-typescript-to-existing-react-project

rissois replied:
I think you highlighted / are looking at the wrong line above, `previewDimensions` is a boolean while `fileDimensions` is never used as a prop.

rissois commented on react-ui/src/components/SubmissionRow.jsx:
```diff
@@ -228,6 +229,7 @@ const updateSubmissions = (currentData, response) => {
 function SubmissionRow({
   moderator,
   submission: {
+    backgroundColor = '#000000',
```
I'm not sure what transparent will do.... We always displayed the flags with a black background

rissois replied:
Ok mod reviewSubmissions now returns the background color. And you are correct about `SubmissionRow`, I made that change as well but kept `EntryModal` to black.

rissois commented on react-ui/src/common/types.js:
```diff
@@ -35,6 +35,7 @@ const types = {
   AppBar: shape({ ...objects.AppBar }),
   color,
   submission: shape({
+    backgroundColor: oneOf(['#000000', '#FFFFFF', '#4b91e3']),
```
Ok, I made the table. The colors are still hardcoded in a few places, you'll have to let me know if these are acceptable: 
1. submission propTypes in `types.js`
2. Default `backgroundColor` as `#000000` in `SubmissionForm`

```
CREATE TABLE background_colors (
    color VARCHAR(7) -- Assuming colors are stored in HEX format, adjust the data type and length as needed
);

INSERT INTO background_colors (color) VALUES
    ('#000000'),
    ('#FFFFFF'),
    ('#4B91E3');
```

heshammourad replied:
I think this looks good, though I haven't had a chance to test it out yet.

heshammourad replied:
Yup, looks good!

rissois commented:
Ok, latest commit summary:

MAJOR CHANGES
1. `env.js` variables were adjusted according to [this comment](https://github.com/heshammourad/vexillology-contests/pull/277#discussion_r1398510617), but note that the technique only worked for`server` and not `react-ui`
2. Background colors is now a separate table that can be created as below:

```
CREATE TABLE background_colors (
    color VARCHAR(7) -- Assuming colors are stored in HEX format, adjust the data type and length as needed
);

INSERT INTO background_colors (color) VALUES
    ('#000000'),
    ('#FFFFFF'),
    ('#4B91E3');
```

MINOR CHANGES
1. `previewDescription` propType. Please check [this thread](https://github.com/heshammourad/vexillology-contests/pull/277#discussion_r1390514002) to make sure that's all good.
2. `EntryModal` key brought down to `default`, and `handleKeyUp` was added to dependencies
3. `PORTAL_` messages

NOT YET ADDRESSED:
1. Dependency cycling is because `useSwrAuth` is not in the `/common` folder. This will be addressed in the final commit when we clean up the structure.
5. Casting timestamps, will report back soon with content of `LOG=debug`
6. Commented out code for upsert. Will either fix it or delete it.
7. Testing keyed entry in staging.

rissois commented:
`LOG = 'debug'` output:

```
9:48:40 PM web.1 |  {"label":"DB","level":"debug","message":"QUERY: update \"contests\" as t set \"name\"=v.\"name\",\"date\"=v.\"date\",\"env_level\"=v.\"env_level\",\"year_end\"=v.\"year_end\",\"winners_thread_id\"=v.\"winners_thread_id\",\"valid_reddit_id\"=v.\"valid_reddit_id\",\"submission_start\"=v.\"submission_start\",\"submission_end\"=v.\"submission_end\",\"vote_start\"=v.\"vote_start\",\"vote_end\"=v.\"vote_end\",\"subtext\"=v.\"subtext\",\"local_voting\"=v.\"local_voting\",\"prompt\"=v.\"prompt\" from (v
9:48:40 PM web.1 |  >  alues('dev','voting','2023-11-26','dev',false,null,false,'2018-10-27T04:48:40.787Z','2018-11-27T05:48:40.787Z','2018-10-27T04:48:40.787Z2','2028-12-27T05:48:40.787Z2','Dev subtext',true,'This is the prompt for the development contest.')) as v(\"id\",\"name\",\"date\",\"env_level\",\"year_end\",\"winners_thread_id\",\"valid_reddit_id\",\"submission_start\",\"submission_end\",\"vote_start\",\"vote_end\",\"subtext\",\"local_voting\",\"prompt\") WHERE v.id = t.id"}
9:48:40 PM web.1 |  {"label":"DB","level":"error","message":"error: column \"date\" is of type date but expression is of type text"}
9:48:40 PM web.1 |  {"label":"API/DEV","level":"error","message":"Error (re)setting /vote: error: column \"date\" is of type date but expression is of type text"}
9:48:40 PM web.1 |  {"label":"REDDIT","level":"debug","message":"Getting username"}
9:48:40 PM web.1 |  {"label":"REDDIT","level":"debug","message":"Getting username"}
9:48:40 PM web.1 |  {"label":"REDDIT","level":"debug","message":"Getting username"}
9:48:40 PM web.1 |  {"label":"REDDIT","level":"debug","message":"Getting username"}
9:48:40 PM web.1 |  {"label":"REDDIT","level":"debug","message":"Getting username"}
9:48:40 PM web.1 |  {"label":"REDDIT","level":"debug","message":"Getting username"}
9:48:41 PM web.1 |  {"label":"REDDIT","level":"debug","message":"Retrieved username: rissois"}
9:48:41 PM web.1 |  {"label":"DB","level":"debug","message":"QUERY: SELECT\n         date,\n         local_voting,\n         name,\n         submission_end,\n         submission_start,\n         subtext,\n         valid_reddit_id,\n         winners_thread_id\n       FROM contests\n       WHERE id = 'dev' AND env_level >= 'dev'"}
9:48:41 PM web.1 |  {"label":"DB","level":"debug","message":"Camelizing column names of: [{\"date\":\"2023-11-21T08:00:00.000Z\",\"local_voting\":true,\"name\":\"submission\",\"submission_end\":\"2028-12-22T04:16:43.788Z\",\"submission_start\":\"2018-10-22T03:16:43.788Z\",\"subtext\":\"Dev subtext\",\"valid_reddit_id\":false,\"winners_thread_id\":null}]"}
9:48:41 PM web.1 |  {"label":"UTIL","level":"debug","message":"Camelizing [{\"date\":\"2023-11-21T08:00:00.000Z\",\"local_voting\":true,\"name\":\"submission\",\"submission_end\":\"2028-12-22T04:16:43.788Z\",\"submission_start\":\"2018-10-22T03:16:43.788Z\",\"subtext\":\"Dev subtext\",\"valid_reddit_id\":false,\"winners_thread_id\":null}]"}
9:48:41 PM web.1 |  {"label":"UTIL","level":"debug","message":"Camelized to [{\"date\":\"2023-11-21T08:00:00.000Z\",\"name\":\"submission\",\"subtext\":\"Dev subtext\",\"localVoting\":true,\"submissionEnd\":\"2028-12-22T04:16:43.788Z\",\"submissionStart\":\"2018-10-22T03:16:43.788Z\",\"validRedditId\":false,\"winnersThreadId\":null}]"}
9:48:41 PM web.1 |  {"label":"DB","level":"debug","message":"RECEIVE: [{\"date\":\"2023-11-21T08:00:00.000Z\",\"name\":\"submission\",\"subtext\":\"Dev subtext\",\"localVoting\":true,\"submissionEnd\":\"2028-12-22T04:16:43.788Z\",\"submissionStart\":\"2018-10-22T03:16:43.788Z\",\"validRedditId\":false,\"winnersThreadId\":null}]"}
9:48:41 PM web.1 |  {"label":"DB","level":"debug","message":"QUERY: SELECT vote_start, vote_end, now() FROM contests WHERE id = 'dev'"}
9:48:41 PM web.1 |  {"label":"DB","level":"debug","message":"Camelizing column names of: [{\"vote_start\":\"2028-11-22T06:16:43.788Z\",\"vote_end\":\"2028-12-22T06:16:43.788Z\",\"now\":\"2023-11-27T05:48:41.290Z\"}]"}
9:48:41 PM web.1 |  {"label":"UTIL","level":"debug","message":"Camelizing [{\"vote_start\":\"2028-11-22T06:16:43.788Z\",\"vote_end\":\"2028-12-22T06:16:43.788Z\",\"now\":\"2023-11-27T05:48:41.290Z\"}]"}
9:48:41 PM web.1 |  {"label":"UTIL","level":"debug","message":"Camelized to [{\"now\":\"2023-11-27T05:48:41.290Z\",\"voteStart\":\"2028-11-22T06:16:43.788Z\",\"voteEnd\":\"2028-12-22T06:16:43.788Z\"}]"}
9:48:41 PM web.1 |  {"label":"DB","level":"debug","message":"RECEIVE: [{\"now\":\"2023-11-27T05:48:41.290Z\",\"voteStart\":\"2028-11-22T06:16:43.788Z\",\"voteEnd\":\"2028-12-22T06:16:43.788Z\"}]"}
9:48:41 PM web.1 |  {"label":"REDDIT","level":"debug","message":"Getting username"}
9:48:41 PM web.1 |  {"label":"REDDIT","level":"debug","message":"Retrieved username: rissois"}
9:48:41 PM web.1 |  {"label":"DB","level":"debug","message":"QUERY: SELECT\n         date,\n         local_voting,\n         name,\n         submission_end,\n         submission_start,\n         subtext,\n         valid_reddit_id,\n         winners_thread_id\n       FROM contests\n       WHERE id = 'dev' AND env_level >= 'dev'"}
9:48:41 PM web.1 |  {"label":"DB","level":"debug","message":"Camelizing column names of: [{\"date\":\"2023-11-21T08:00:00.000Z\",\"local_voting\":true,\"name\":\"submission\",\"submission_end\":\"2028-12-22T04:16:43.788Z\",\"submission_start\":\"2018-10-22T03:16:43.788Z\",\"subtext\":\"Dev subtext\",\"valid_reddit_id\":false,\"winners_thread_id\":null}]"}
9:48:41 PM web.1 |  {"label":"UTIL","level":"debug","message":"Camelizing [{\"date\":\"2023-11-21T08:00:00.000Z\",\"local_voting\":true,\"name\":\"submission\",\"submission_end\":\"2028-12-22T04:16:43.788Z\",\"submission_start\":\"2018-10-22T03:16:43.788Z\",\"subtext\":\"Dev subtext\",\"valid_reddit_id\":false,\"winners_thread_id\":null}]"}
9:48:41 PM web.1 |  {"label":"UTIL","level":"debug","message":"Camelized to [{\"date\":\"2023-11-21T08:00:00.000Z\",\"name\":\"submission\",\"subtext\":\"Dev subtext\",\"localVoting\":true,\"submissionEnd\":\"2028-12-22T04:16:43.788Z\",\"submissionStart\":\"2018-10-22T03:16:43.788Z\",\"validRedditId\":false,\"winnersThreadId\":null}]"}
9:48:41 PM web.1 |  {"label":"DB","level":"debug","message":"RECEIVE: [{\"date\":\"2023-11-21T08:00:00.000Z\",\"name\":\"submission\",\"subtext\":\"Dev subtext\",\"localVoting\":true,\"submissionEnd\":\"2028-12-22T04:16:43.788Z\",\"submissionStart\":\"2018-10-22T03:16:43.788Z\",\"validRedditId\":false,\"winnersThreadId\":null}]"}
9:48:41 PM web.1 |  {"label":"DB","level":"debug","message":"QUERY: SELECT vote_start, vote_end, now() FROM contests WHERE id = 'dev'"}
9:48:41 PM web.1 |  {"label":"DB","level":"debug","message":"Camelizing column names of: [{\"vote_start\":\"2028-11-22T06:16:43.788Z\",\"vote_end\":\"2028-12-22T06:16:43.788Z\",\"now\":\"2023-11-27T05:48:41.328Z\"}]"}
9:48:41 PM web.1 |  {"label":"UTIL","level":"debug","message":"Camelizing [{\"vote_start\":\"2028-11-22T06:16:43.788Z\",\"vote_end\":\"2028-12-22T06:16:43.788Z\",\"now\":\"2023-11-27T05:48:41.328Z\"}]"}
9:48:41 PM web.1 |  {"label":"UTIL","level":"debug","message":"Camelized to [{\"now\":\"2023-11-27T05:48:41.328Z\",\"voteStart\":\"2028-11-22T06:16:43.788Z\",\"voteEnd\":\"2028-12-22T06:16:43.788Z\"}]"}
9:48:41 PM web.1 |  {"label":"DB","level":"debug","message":"RECEIVE: [{\"now\":\"2023-11-27T05:48:41.328Z\",\"voteStart\":\"2028-11-22T06:16:43.788Z\",\"voteEnd\":\"2028-12-22T06:16:43.788Z\"}]"}
9:48:41 PM web.1 |  {"label":"REDDIT","level":"debug","message":"Retrieved username: rissois"}
9:48:41 PM web.1 |  {"label":"REDDIT","level":"debug","message":"Retrieved username: rissois"}
9:48:41 PM web.1 |  {"label":"DB","level":"debug","message":"QUERY: SELECT\n         date,\n         local_voting,\n         name,\n         submission_end,\n         submission_start,\n         subtext,\n         valid_reddit_id,\n         winners_thread_id\n       FROM contests\n       WHERE id = 'dev' AND env_level >= 'dev'"}
9:48:41 PM web.1 |  {"label":"DB","level":"debug","message":"QUERY: SELECT\n         date,\n         local_voting,\n         name,\n         submission_end,\n         submission_start,\n         subtext,\n         valid_reddit_id,\n         winners_thread_id\n       FROM contests\n       WHERE id = 'dev' AND env_level >= 'dev'"}
9:48:41 PM web.1 |  {"label":"DB","level":"debug","message":"Camelizing column names of: [{\"date\":\"2023-11-21T08:00:00.000Z\",\"local_voting\":true,\"name\":\"submission\",\"submission_end\":\"2028-12-22T04:16:43.788Z\",\"submission_start\":\"2018-10-22T03:16:43.788Z\",\"subtext\":\"Dev subtext\",\"valid_reddit_id\":false,\"winners_thread_id\":null}]"}
9:48:41 PM web.1 |  {"label":"UTIL","level":"debug","message":"Camelizing [{\"date\":\"2023-11-21T08:00:00.000Z\",\"local_voting\":true,\"name\":\"submission\",\"submission_end\":\"2028-12-22T04:16:43.788Z\",\"submission_start\":\"2018-10-22T03:16:43.788Z\",\"subtext\":\"Dev subtext\",\"valid_reddit_id\":false,\"winners_thread_id\":null}]"}
9:48:41 PM web.1 |  {"label":"UTIL","level":"debug","message":"Camelized to [{\"date\":\"2023-11-21T08:00:00.000Z\",\"name\":\"submission\",\"subtext\":\"Dev subtext\",\"localVoting\":true,\"submissionEnd\":\"2028-12-22T04:16:43.788Z\",\"submissionStart\":\"2018-10-22T03:16:43.788Z\",\"validRedditId\":false,\"winnersThreadId\":null}]"}
9:48:41 PM web.1 |  {"label":"DB","level":"debug","message":"RECEIVE: [{\"date\":\"2023-11-21T08:00:00.000Z\",\"name\":\"submission\",\"subtext\":\"Dev subtext\",\"localVoting\":true,\"submissionEnd\":\"2028-12-22T04:16:43.788Z\",\"submissionStart\":\"2018-10-22T03:16:43.788Z\",\"validRedditId\":false,\"winnersThreadId\":null}]"}
9:48:41 PM web.1 |  {"label":"DB","level":"debug","message":"QUERY: SELECT vote_start, vote_end, now() FROM contests WHERE id = 'dev'"}
9:48:41 PM web.1 |  {"label":"DB","level":"debug","message":"Camelizing column names of: [{\"date\":\"2023-11-21T08:00:00.000Z\",\"local_voting\":true,\"name\":\"submission\",\"submission_end\":\"2028-12-22T04:16:43.788Z\",\"submission_start\":\"2018-10-22T03:16:43.788Z\",\"subtext\":\"Dev subtext\",\"valid_reddit_id\":false,\"winners_thread_id\":null}]"}
9:48:41 PM web.1 |  {"label":"UTIL","level":"debug","message":"Camelizing [{\"date\":\"2023-11-21T08:00:00.000Z\",\"local_voting\":true,\"name\":\"submission\",\"submission_end\":\"2028-12-22T04:16:43.788Z\",\"submission_start\":\"2018-10-22T03:16:43.788Z\",\"subtext\":\"Dev subtext\",\"valid_reddit_id\":false,\"winners_thread_id\":null}]"}
9:48:41 PM web.1 |  {"label":"UTIL","level":"debug","message":"Camelized to [{\"date\":\"2023-11-21T08:00:00.000Z\",\"name\":\"submission\",\"subtext\":\"Dev subtext\",\"localVoting\":true,\"submissionEnd\":\"2028-12-22T04:16:43.788Z\",\"submissionStart\":\"2018-10-22T03:16:43.788Z\",\"validRedditId\":false,\"winnersThreadId\":null}]"}
9:48:41 PM web.1 |  {"label":"DB","level":"debug","message":"RECEIVE: [{\"date\":\"2023-11-21T08:00:00.000Z\",\"name\":\"submission\",\"subtext\":\"Dev subtext\",\"localVoting\":true,\"submissionEnd\":\"2028-12-22T04:16:43.788Z\",\"submissionStart\":\"2018-10-22T03:16:43.788Z\",\"validRedditId\":false,\"winnersThreadId\":null}]"}
9:48:41 PM web.1 |  {"label":"DB","level":"debug","message":"QUERY: SELECT vote_start, vote_end, now() FROM contests WHERE id = 'dev'"}
9:48:41 PM web.1 |  {"label":"DB","level":"debug","message":"Camelizing column names of: [{\"vote_start\":\"2028-11-22T06:16:43.788Z\",\"vote_end\":\"2028-12-22T06:16:43.788Z\",\"now\":\"2023-11-27T05:48:41.354Z\"}]"}
9:48:41 PM web.1 |  {"label":"UTIL","level":"debug","message":"Camelizing [{\"vote_start\":\"2028-11-22T06:16:43.788Z\",\"vote_end\":\"2028-12-22T06:16:43.788Z\",\"now\":\"2023-11-27T05:48:41.354Z\"}]"}
9:48:41 PM web.1 |  {"label":"UTIL","level":"debug","message":"Camelized to [{\"now\":\"2023-11-27T05:48:41.354Z\",\"voteStart\":\"2028-11-22T06:16:43.788Z\",\"voteEnd\":\"2028-12-22T06:16:43.788Z\"}]"}
9:48:41 PM web.1 |  {"label":"DB","level":"debug","message":"RECEIVE: [{\"now\":\"2023-11-27T05:48:41.354Z\",\"voteStart\":\"2028-11-22T06:16:43.788Z\",\"voteEnd\":\"2028-12-22T06:16:43.788Z\"}]"}
9:48:41 PM web.1 |  {"label":"DB","level":"debug","message":"Camelizing column names of: [{\"vote_start\":\"2028-11-22T06:16:43.788Z\",\"vote_end\":\"2028-12-22T06:16:43.788Z\",\"now\":\"2023-11-27T05:48:41.355Z\"}]"}
9:48:41 PM web.1 |  {"label":"UTIL","level":"debug","message":"Camelizing [{\"vote_start\":\"2028-11-22T06:16:43.788Z\",\"vote_end\":\"2028-12-22T06:16:43.788Z\",\"now\":\"2023-11-27T05:48:41.355Z\"}]"}
9:48:41 PM web.1 |  {"label":"UTIL","level":"debug","message":"Camelized to [{\"now\":\"2023-11-27T05:48:41.355Z\",\"voteStart\":\"2028-11-22T06:16:43.788Z\",\"voteEnd\":\"2028-12-22T06:16:43.788Z\"}]"}
9:48:41 PM web.1 |  {"label":"DB","level":"debug","message":"RECEIVE: [{\"now\":\"2023-11-27T05:48:41.355Z\",\"voteStart\":\"2028-11-22T06:16:43.788Z\",\"voteEnd\":\"2028-12-22T06:16:43.788Z\"}]"}
9:48:41 PM web.1 |  {"label":"REDDIT","level":"debug","message":"Retrieved username: rissois"}
9:48:41 PM web.1 |  {"label":"REDDIT","level":"debug","message":"Retrieved username: rissois"}
9:48:41 PM web.1 |  {"label":"DB","level":"debug","message":"QUERY: SELECT\n         date,\n         local_voting,\n         name,\n         submission_end,\n         submission_start,\n         subtext,\n         valid_reddit_id,\n         winners_thread_id\n       FROM contests\n       WHERE id = 'dev' AND env_level >= 'dev'"}
9:48:41 PM web.1 |  {"label":"DB","level":"debug","message":"QUERY: SELECT\n         date,\n         local_voting,\n         name,\n         submission_end,\n         submission_start,\n         subtext,\n         valid_reddit_id,\n         winners_thread_id\n       FROM contests\n       WHERE id = 'dev' AND env_level >= 'dev'"}
9:48:41 PM web.1 |  {"label":"DB","level":"debug","message":"Camelizing column names of: [{\"date\":\"2023-11-21T08:00:00.000Z\",\"local_voting\":true,\"name\":\"submission\",\"submission_end\":\"2028-12-22T04:16:43.788Z\",\"submission_start\":\"2018-10-22T03:16:43.788Z\",\"subtext\":\"Dev subtext\",\"valid_reddit_id\":false,\"winners_thread_id\":null}]"}
9:48:41 PM web.1 |  {"label":"UTIL","level":"debug","message":"Camelizing [{\"date\":\"2023-11-21T08:00:00.000Z\",\"local_voting\":true,\"name\":\"submission\",\"submission_end\":\"2028-12-22T04:16:43.788Z\",\"submission_start\":\"2018-10-22T03:16:43.788Z\",\"subtext\":\"Dev subtext\",\"valid_reddit_id\":false,\"winners_thread_id\":null}]"}
9:48:41 PM web.1 |  {"label":"UTIL","level":"debug","message":"Camelized to [{\"date\":\"2023-11-21T08:00:00.000Z\",\"name\":\"submission\",\"subtext\":\"Dev subtext\",\"localVoting\":true,\"submissionEnd\":\"2028-12-22T04:16:43.788Z\",\"submissionStart\":\"2018-10-22T03:16:43.788Z\",\"validRedditId\":false,\"winnersThreadId\":null}]"}
9:48:41 PM web.1 |  {"label":"DB","level":"debug","message":"RECEIVE: [{\"date\":\"2023-11-21T08:00:00.000Z\",\"name\":\"submission\",\"subtext\":\"Dev subtext\",\"localVoting\":true,\"submissionEnd\":\"2028-12-22T04:16:43.788Z\",\"submissionStart\":\"2018-10-22T03:16:43.788Z\",\"validRedditId\":false,\"winnersThreadId\":null}]"}
9:48:41 PM web.1 |  {"label":"DB","level":"debug","message":"QUERY: SELECT vote_start, vote_end, now() FROM contests WHERE id = 'dev'"}
9:48:41 PM web.1 |  {"label":"DB","level":"debug","message":"Camelizing column names of: [{\"date\":\"2023-11-21T08:00:00.000Z\",\"local_voting\":true,\"name\":\"submission\",\"submission_end\":\"2028-12-22T04:16:43.788Z\",\"submission_start\":\"2018-10-22T03:16:43.788Z\",\"subtext\":\"Dev subtext\",\"valid_reddit_id\":false,\"winners_thread_id\":null}]"}
9:48:41 PM web.1 |  {"label":"UTIL","level":"debug","message":"Camelizing [{\"date\":\"2023-11-21T08:00:00.000Z\",\"local_voting\":true,\"name\":\"submission\",\"submission_end\":\"2028-12-22T04:16:43.788Z\",\"submission_start\":\"2018-10-22T03:16:43.788Z\",\"subtext\":\"Dev subtext\",\"valid_reddit_id\":false,\"winners_thread_id\":null}]"}
9:48:41 PM web.1 |  {"label":"UTIL","level":"debug","message":"Camelized to [{\"date\":\"2023-11-21T08:00:00.000Z\",\"name\":\"submission\",\"subtext\":\"Dev subtext\",\"localVoting\":true,\"submissionEnd\":\"2028-12-22T04:16:43.788Z\",\"submissionStart\":\"2018-10-22T03:16:43.788Z\",\"validRedditId\":false,\"winnersThreadId\":null}]"}
9:48:41 PM web.1 |  {"label":"DB","level":"debug","message":"RECEIVE: [{\"date\":\"2023-11-21T08:00:00.000Z\",\"name\":\"submission\",\"subtext\":\"Dev subtext\",\"localVoting\":true,\"submissionEnd\":\"2028-12-22T04:16:43.788Z\",\"submissionStart\":\"2018-10-22T03:16:43.788Z\",\"validRedditId\":false,\"winnersThreadId\":null}]"}
9:48:41 PM web.1 |  {"label":"DB","level":"debug","message":"QUERY: SELECT vote_start, vote_end, now() FROM contests WHERE id = 'dev'"}
9:48:41 PM web.1 |  {"label":"DB","level":"debug","message":"Camelizing column names of: [{\"vote_start\":\"2028-11-22T06:16:43.788Z\",\"vote_end\":\"2028-12-22T06:16:43.788Z\",\"now\":\"2023-11-27T05:48:41.401Z\"}]"}
9:48:41 PM web.1 |  {"label":"UTIL","level":"debug","message":"Camelizing [{\"vote_start\":\"2028-11-22T06:16:43.788Z\",\"vote_end\":\"2028-12-22T06:16:43.788Z\",\"now\":\"2023-11-27T05:48:41.401Z\"}]"}
9:48:41 PM web.1 |  {"label":"UTIL","level":"debug","message":"Camelized to [{\"now\":\"2023-11-27T05:48:41.401Z\",\"voteStart\":\"2028-11-22T06:16:43.788Z\",\"voteEnd\":\"2028-12-22T06:16:43.788Z\"}]"}
9:48:41 PM web.1 |  {"label":"DB","level":"debug","message":"RECEIVE: [{\"now\":\"2023-11-27T05:48:41.401Z\",\"voteStart\":\"2028-11-22T06:16:43.788Z\",\"voteEnd\":\"2028-12-22T06:16:43.788Z\"}]"}
9:48:41 PM web.1 |  {"label":"DB","level":"debug","message":"Camelizing column names of: [{\"vote_start\":\"2028-11-22T06:16:43.788Z\",\"vote_end\":\"2028-12-22T06:16:43.788Z\",\"now\":\"2023-11-27T05:48:41.402Z\"}]"}
9:48:41 PM web.1 |  {"label":"UTIL","level":"debug","message":"Camelizing [{\"vote_start\":\"2028-11-22T06:16:43.788Z\",\"vote_end\":\"2028-12-22T06:16:43.788Z\",\"now\":\"2023-11-27T05:48:41.402Z\"}]"}
9:48:41 PM web.1 |  {"label":"UTIL","level":"debug","message":"Camelized to [{\"now\":\"2023-11-27T05:48:41.402Z\",\"voteStart\":\"2028-11-22T06:16:43.788Z\",\"voteEnd\":\"2028-12-22T06:16:43.788Z\"}]"}
9:48:41 PM web.1 |  {"label":"DB","level":"debug","message":"RECEIVE: [{\"now\":\"2023-11-27T05:48:41.402Z\",\"voteStart\":\"2028-11-22T06:16:43.788Z\",\"voteEnd\":\"2028-12-22T06:16:43.788Z\"}]"}
9:48:41 PM web.1 |  {"label":"REDDIT","level":"debug","message":"Retrieved username: rissois"}
9:48:41 PM web.1 |  {"label":"DB","level":"debug","message":"QUERY: SELECT\n         date,\n         local_voting,\n         name,\n         submission_end,\n         submission_start,\n         subtext,\n         valid_reddit_id,\n         winners_thread_id\n       FROM contests\n       WHERE id = 'dev' AND env_level >= 'dev'"}
9:48:41 PM web.1 |  {"label":"DB","level":"debug","message":"Camelizing column names of: [{\"date\":\"2023-11-21T08:00:00.000Z\",\"local_voting\":true,\"name\":\"submission\",\"submission_end\":\"2028-12-22T04:16:43.788Z\",\"submission_start\":\"2018-10-22T03:16:43.788Z\",\"subtext\":\"Dev subtext\",\"valid_reddit_id\":false,\"winners_thread_id\":null}]"}
9:48:41 PM web.1 |  {"label":"UTIL","level":"debug","message":"Camelizing [{\"date\":\"2023-11-21T08:00:00.000Z\",\"local_voting\":true,\"name\":\"submission\",\"submission_end\":\"2028-12-22T04:16:43.788Z\",\"submission_start\":\"2018-10-22T03:16:43.788Z\",\"subtext\":\"Dev subtext\",\"valid_reddit_id\":false,\"winners_thread_id\":null}]"}
9:48:41 PM web.1 |  {"label":"UTIL","level":"debug","message":"Camelized to [{\"date\":\"2023-11-21T08:00:00.000Z\",\"name\":\"submission\",\"subtext\":\"Dev subtext\",\"localVoting\":true,\"submissionEnd\":\"2028-12-22T04:16:43.788Z\",\"submissionStart\":\"2018-10-22T03:16:43.788Z\",\"validRedditId\":false,\"winnersThreadId\":null}]"}
9:48:41 PM web.1 |  {"label":"DB","level":"debug","message":"RECEIVE: [{\"date\":\"2023-11-21T08:00:00.000Z\",\"name\":\"submission\",\"subtext\":\"Dev subtext\",\"localVoting\":true,\"submissionEnd\":\"2028-12-22T04:16:43.788Z\",\"submissionStart\":\"2018-10-22T03:16:43.788Z\",\"validRedditId\":false,\"winnersThreadId\":null}]"}
9:48:41 PM web.1 |  {"label":"DB","level":"debug","message":"QUERY: SELECT vote_start, vote_end, now() FROM contests WHERE id = 'dev'"}
9:48:41 PM web.1 |  {"label":"DB","level":"debug","message":"Camelizing column names of: [{\"vote_start\":\"2028-11-22T06:16:43.788Z\",\"vote_end\":\"2028-12-22T06:16:43.788Z\",\"now\":\"2023-11-27T05:48:41.758Z\"}]"}
9:48:41 PM web.1 |  {"label":"UTIL","level":"debug","message":"Camelizing [{\"vote_start\":\"2028-11-22T06:16:43.788Z\",\"vote_end\":\"2028-12-22T06:16:43.788Z\",\"now\":\"2023-11-27T05:48:41.758Z\"}]"}
9:48:41 PM web.1 |  {"label":"UTIL","level":"debug","message":"Camelized to [{\"now\":\"2023-11-27T05:48:41.758Z\",\"voteStart\":\"2028-11-22T06:16:43.788Z\",\"voteEnd\":\"2028-12-22T06:16:43.788Z\"}]"}
9:48:41 PM web.1 |  {"label":"DB","level":"debug","message":"RECEIVE: [{\"now\":\"2023-11-27T05:48:41.758Z\",\"voteStart\":\"2028-11-22T06:16:43.788Z\",\"voteEnd\":\"2028-12-22T06:16:43.788Z\"}]"}
```

rissois commented on react-ui/src/common/useVoting.jsx:
```diff
@@ -0,0 +1,110 @@
+    },
+  });
+
+  // ??? Should this be a async and auto-next?
```
Ok sounds good

rissois commented on react-ui/src/common/useVotingStatus.jsx:
```diff
@@ -0,0 +1,57 @@
+
+import useComponentsState from './useComponentsState';
+
+const MAX_DELAY = 2147483647;
```
It's the maximum allowable value for `setTimeout`

heshammourad replied:
Ah, TIL. Does it really make sense to use this?

In the unlikely event that we have a voting window open longer than 24.8 days, and the user actually keeps their browser running and tab focused for that duration, to then fire the end voting code?

Perhaps we need to branch if the delay is longer than 24.8 days and reevaluate this condition after the timeout rather than firing the mutation and marking as expired.

On the other hand, this would be handling quite the edge case that is extremely unlikely to ever need to be accounted for, so maybe just add a comment that this is the max allowable value and call it a day?

rissois replied:
So I originally implemented this because the `demo` contest delay is longer than 24.8 days, and this caused the timeout to fire immediately. I went with the comment approach because as you pointed out, seems better to handle the edge case than restrict everything else to avoid triggering an edge case.

rissois commented on react-ui/src/components/AppBarDivided.jsx:
```diff
@@ -22,8 +23,10 @@ function AppBarDivided({
   position,
   right,
 }) {
+  const isModal = accountMenuColor === 'inherit';
```
Yep you're right, I hadn't caught the code extended into AppBarDivided. I guess this is exactly why we switched to a prop.

rissois commented on react-ui/src/components/EntryDescriptionDrawer.jsx:
```diff
@@ -26,7 +29,7 @@ import RedditMarkdown from './RedditMarkdown';
 import VotingSlider from './VotingSlider';
 
-const useStyles = makeStyles({
+const useStyles = makeStyles(() => ({
```
Must have been playing around with something, reverted back.

rissois commented on react-ui/src/pages/contest/Contest.jsx:
```diff
@@ -87,37 +80,10 @@ function Contest() {
   useEffect(() => {
     // Clear cache if the voting window is still closed to force fetch again on next visit
```
There is still a built-in cache for `swr`, so yes you still need mutate when the voting window closes to request swr to refetch the data.

rissois commented on react-ui/src/pages/contest/ContestAppBarMain.jsx:
```diff
@@ -37,7 +37,6 @@ export default function ContestAppBarMain({ handleVotingExpired, handleReload, c
         <Box display="inline-flex" paddingLeft={1.5}>
           <Countdown
             endDate={voteEndDate}
-            handleExpiry={handleVotingExpired}
```
`handleExpiry` is still required for `Submission.jsx`

rissois commented on react-ui/src/pages/contest/ContestAppBarMain.jsx:
```diff
@@ -24,11 +26,9 @@ export default function ContestAppBarMain({ handleVotingExpired, handleReload, c
 
   if (!name) {
-    return null;
+    return <RouterLinkIconButton state={{ date }} to={backLink} />;
```
Hm yeah, I sort of forget what I introduced `!name`, but I think it was because I felt name *had* to be there or we had bigger problems to investigate. I took it out.

heshammourad replied:
I thought `!name` was there from when I added the code. It was generally a substitute for "have fetched data" since name was always going to be included, but it's cleaner now that we always include the back button.

rissois replied:
Ah ok, sounds good thought it might have been from my refactoring

rissois commented on react-ui/src/pages/submission/Submission.jsx:
```diff
@@ -135,277 +38,57 @@ const useStyles = makeStyles((theme) => ({
+      <InternalLink to={`/contests/${contestId}`}>here</InternalLink>
+      &nbsp;to view entries.
+    </div>
+  );
```
Back to inline then

heshammourad commented:
Github makes adding comment threads really difficult.

Here's the code snippet that I have that is updating dates:
```js
const voteData = {
  contest_id: contestId,
  entry_id: entryId,
  username,
  rating,
  last_modified: new Date(),
};
await db.update(
  'votes',
  [voteData],
  [
    '?contest_id',
    '?entry_id',
    '?username',
    'rating',
    { name: 'last_modified', cast: 'timestamp' },
  ],
);
```
The logged output has `::timestamp` after the date in the values, which is missing from your log.
```
{"label":"DB","level":"debug","message":"QUERY: update \"votes\" as t set \"rating\"=v.\"rating\",\"last_modified\"=v.\"last_modified\" from (values('zt8ghv','UMbMEGu','heshammourad',1,'2023-11-26T22:48:36.385-08:00'::timestamp)) as v(\"contest_id\",\"entry_id\",\"username\",\"rating\",\"last_modified\") WHERE v.contest_id = t.contest_id AND v.entry_id = t.entry_id AND v.username = t.username"}
```

Can you also include the `db.update` line you have?

heshammourad commented:
I think you may still have `.toISOString()` on each of those lines from our earlier attempt to solve this. The code you linked to is looking for a date to decide to cast to timestamp, so if you remove the `toISOString`s, it should hopefully work. Give it a try when you have some time.

heshammourad commented on react-ui/src/pages/submission/SubmissionForm.jsx:
```diff
@@ -0,0 +1,618 @@
+            <Typography variant="caption">Background color</Typography>
+            <Stack direction="row" spacing={2}>
+              {backgroundColors.map((color) => (
+                <Box
```
This should probably be a button, so we can get the cursor changes, etc.

heshammourad commented on react-ui/src/pages/entry/EntryModal.jsx:
```diff
@@ -179,9 +177,16 @@ function EntryModal() {
+          changeRating(parseInt(key, 10), entry?.rating, true);
+          indexChange = 1;
+        }
         return;
```
If you want the `indexChange` to be processed, you need a `break` instead of `returning`. I'm a bit worried that errors would be hard to discern and votes might be missed.

My suggestion is change the `changeRating` and `clearRating` functions in `useVoting.jsx` to `async`. And then here:
```js
  const handleKeyedVote = async (rating) => {
    await changeRating(rating, entry?.rating, true);
    handleNavigate(1);
  };

    switch (key) {
      // existing switch cases
      default:
        if (key >= '0' && key <= '5') {
          handleKeyedVote(parseInt(key, 10));
        }
        return; // Note that I've kept return here, because I don't want to handle navigate here and I can't await in an event handler (that I know of).
    }
```

rissois replied:
Ah ok, so the main issue is that voting often seems to be a little slow / delayed. So there's three options:
1. Auto-navigate asynchronously, but a) the snackbar will appear on the next flag and b) you can still vote for the next flag, which if performed before the previous call completes seem to cancel it out... not sure why
2. Auto-navigate synchronously, but then the delay could lead to key spamming
3. Don't auto-navigate, which may be the best intermediate step for now and we can revisit the asynchronous failure later.

heshammourad replied:
I'm fine with 3 for now since it looks like there's a couple of issues to work through.

heshammourad commented:
> Ok I think I finally got Postgres and PG-Promise to work with each other. Once you've been able to look at this, we can address the following:
> 
>     Dependency cycling is because useSwrAuth is not in the /common folder. This will be addressed in the final commit when we clean up the structure.
>     Commented out code for upsert. I'm honestly going to delete this rather than touch PG-Promise again for a while.
>     Testing keyed entry in staging.

I tested the keyed entry and overall looks good. I added a comment near that logic with a suggestion on how to handle auto-navigation (which doesn't currently work). I'm fine with making the suggested change or leaving it as-is with no auto-next functionality.

I also created #289 which can be addressed later, but the entry modal app bar looks a bit off when it doesn't match the background color.

rissois commented on react-ui/src/pages/entry/EntryModal.jsx:
```diff
@@ -179,9 +177,16 @@ function EntryModal() {
+        if (key >= '0' && key <= '5') {
+          // ??? I would love it if we could track whether votes mobile or desktop
+          // ??? and if desktop, how many were keyed
+          // https://stackoverflow.com/questions/39435395/reactjs-how-to-determine-if-the-application-is-being-viewed-on-mobile-or-deskto
```
I love that actually, added inline and removed old comments.

rissois commented:
LATEST CHANGES:
- Converted background chips into buttons
- Adjusted appBar color to match background color
- Introduced new system for `env.js` (see `docs/README.md` for details)

TO DO
- Test `upsert` (or move to an issue for future implementation)
- Merge `api` and `utils` into `data` directory. Move `useSwrAuth` into `common` directory
- Test `is_keyed` and `is_desktop` voting on mobile

heshammourad reviewed (approved):
Went through and tested the things and looked at the open comments. I added a few comments for minor things in #294, they're not blocking, so we can go ahead and merge this and fix them in that PR.

I also used the dev bar functionality for the first time to create a submission contest and test out the color buttons. Really cool how it works, look forward to using it more in the future! 🎉
